### PR TITLE
Initial Sculpting by shortening bow tie pulls

### PR DESCRIPTION
**Pull request by geralddejong**

### CODE REVIEW

File: src/application.rs [hunk 1]
```diff
@@ -156,10 +157,13 @@ pub fn run() {
...
-                let message = app.scene.update(&app.graphics, app.gui.controls(), experiment.fabric());
-                if let Some(message) = message {
-                    app.gui.change_state(message)
-                }
+                let strain_view =
+                    if app.gui.controls().show_strain() {
+                        Some(app.gui.controls().strain_view())
+                    } else {
+                        None
+                    };
```
Comment (mitchelljustin) [inline src/application.rs:165]:
```suggestion
                let strain_view =
                    app.gui.controls().show_strain().then(|| app.gui.controls().strain_view());
;
;
```

File: src/controls/mod.rs [hunk 2]
```diff
@@ -223,30 +226,43 @@ impl Default for ControlState {
...
+            show_strain: false,
             frame_rate: 0.0,
             action_queue: RefCell::new(Vec::new()),
         }
     }
 }
 
+pub struct StrainView {
+    pub threshold: f32,
+    pub material: usize,
```
Comment (mitchelljustin) [inline src/controls/mod.rs:245]: Do you plan on varying the material in the future?

Comment (geralddejong) [reply src/controls/mod.rs:245]: Not sure, there's only one special one now.

File: src/fabric/vulcanize.rs [hunk 3]
```diff
@@ -1,8 +1,9 @@
 use std::collections::{HashMap, HashSet};
 use cgmath::{MetricSpace, Point3};
+use chrono::format::Fixed;
```
Comment (mitchelljustin) [inline src/fabric/vulcanize.rs:3]:
```suggestion
```
Probably a mistake

File: src/fabric/vulcanize.rs [hunk 4]
```diff
@@ -16,21 +17,31 @@ const BOW_TIE_MATERIAL: Material = Material {
...
+    }
+
+    pub fn shorten_pulls(&mut self, strain_threshold: f32) {
+        for interval in self.intervals.values_mut() {
+            if interval.material != Self::BOW_TIE_MATERIAL_INDEX {
+                continue;
+            }
+            if interval.strain > strain_threshold {
+                interval.span = match interval.span {
+                    Span::Fixed { length } => Span::Fixed { length: length * 0.95 },
```
Comment (mitchelljustin) [inline src/fabric/vulcanize.rs:41]: Extract `0.95` to a const

File: src/fabric/vulcanize.rs [hunk 5]
```diff
@@ -16,21 +16,31 @@ const BOW_TIE_MATERIAL: Material = Material {
...
+
+    pub fn shorten_pulls(&mut self, strain_threshold: f32, shortening: f32) {
+        for interval in self.intervals.values_mut() {
+            if interval.material != Self::BOW_TIE_MATERIAL_INDEX {
+                continue;
+            }
+            if interval.strain > strain_threshold {
+                interval.span = match interval.span {
+                    Span::Fixed { length } => Span::Fixed { length: length * shortening },
+                    _ => panic!()
```
Comment (mitchelljustin) [inline src/fabric/vulcanize.rs:41]:
```suggestion
                    _ => unreachable!()
```

**Review by mitchelljustin — APPROVED** [review]

**Pull request merged**